src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -75,49 +127,107 @@ object RockSlicer {
+              current +: unique
+            } else {
+              unique
+            }
+          }
+        if (allFaces.diff(nonSharedFaces).isEmpty) {
+          Some(Block(origin, nonSharedFaces))
+        } else {
+          None
+        }
+      } else {
+        None
+      }
+    }).flatten
+
+    // Remove redundant faces and remove duplicates
+    val joinedBlocks = (reconstructedBlocks map { case block @ Block(center, _) =>
+      Block(center, block.nonRedundantFaces)
+    }).filter { block => block.faces.nonEmpty } // Filters blocks that actually aren't adjacent at all,
+                                                // but shared a processor joint
+
+    // Divide blocks into groups with and without processor joints remaining
+    val (remainingBlocks, completedBlocks) = joinedBlocks.partition { block =>
+      block.faces.exists(_.processorJoint)
     }
-  }
 
-  private def loadBalanceIndices(indices: Seq[Int]): Seq[Int] = {
-    def loadBalanceIndicesImpl(queue: mutable.PriorityQueue[Seq[Int]]): Seq[Int] = {
-      if (queue.isEmpty) {
-        Nil
-      } else {
-        val s = queue.dequeue()
-        val (median, left, right) = removeMedian(s)
-        if (left.nonEmpty) {
-          queue.enqueue(left)
+    if (remainingBlocks.nonEmpty) {
+      // Merged blocks still contain some processor joints
+      mergeBlocks(remainingBlocks ++ processorBlocks.tail, completedBlocks ++ mergedBlocks, origin)
+    } else if (processorBlocks.tail.isEmpty) {
+      // All blocks are free of processor joints - check for duplicates then return
+      val mergedBlocksDuplicates = completedBlocks ++ mergedBlocks
+      mergedBlocksDuplicates.foldLeft(Seq[Block]()) { (unique, current) =>
+        if (!unique.exists(current.approximateEquals(_))) {
+          current +: unique
+        } else {
+          unique
         }
-        if (right.nonEmpty) {
-          queue.enqueue(right)
-        }
-        median +: loadBalanceIndicesImpl(queue)
       }
+    } else {
+      // Proceed to next processor block
+      mergeBlocks(processorBlocks.tail, completedBlocks ++ mergedBlocks, origin)
     }
+  }
 
-    def removeMedian[A](s: Seq[A]): (A, Seq[A], Seq[A]) = {
-      val medianIndex = s.length / 2
-      (s(medianIndex), s.take(medianIndex), s.drop(medianIndex + 1))
-    }
+  /**
+    * Compares two input blocks and determines whether they share a processor face
+    * @param block1 First input block
+    * @param block2 Second input block
+    * @return List of processor faces that are shared by the two blocks. Will be 
+    *         empty if they share no faces
+    */
+  def compareProcessorBlocks(block1: Block, block2: Block): Seq[Face] = {
+    val processorFaces1 = block1.faces.filter { face => face.processorJoint }
+    val processorFaces2 = block2.faces.filter { face => face.processorJoint }
+    val faceMatches = 
+      processorFaces1 map { case face1 =>
+        processorFaces2 map { case face2 =>
```

Comment:
I think the primary reason this won't work is because the number of processor joints per block aren't guaranteed to be the same across all blocks, so even when you do a sort how do you do a fair comparison between the faces without some sort of map? I tried implementing a sorting-type solution, but it only works when the number of processor joints per block are the same between the two blocks being compared. 
For example, if you have a single processor joint in one block and three in another, sorting is somewhat pointless since you'll need to check the single joint against each of the other three for a match anyhow.
Do you agree with that?